vfs/xts/xts.go

```diff
@@ -0,0 +1,275 @@
+package xts
+
+import (
+	"encoding/hex"
+	"io"
+
+	"github.com/ncruces/go-sqlite3"
+	"github.com/ncruces/go-sqlite3/internal/util"
+	"github.com/ncruces/go-sqlite3/vfs"
+	"golang.org/x/crypto/xts"
+)
+
+type xtsVFS struct {
+	vfs.VFS
+	init XTSCreator
+}
+
+func (x *xtsVFS) Open(name string, flags vfs.OpenFlag) (vfs.File, vfs.OpenFlag, error) {
+	// notest // OpenFilename is called instead
+	return nil, 0, sqlite3.CANTOPEN
+}
+
+func (x *xtsVFS) OpenFilename(name *vfs.Filename, flags vfs.OpenFlag) (file vfs.File, _ vfs.OpenFlag, err error) {
+	if hf, ok := x.VFS.(vfs.VFSFilename); ok {
+		file, flags, err = hf.OpenFilename(name, flags)
+	} else {
+		file, flags, err = x.VFS.Open(name.String(), flags)
+	}
+
+	// Encrypt everything except super journals and memory files.
+	if err != nil || flags&(vfs.OPEN_SUPER_JOURNAL|vfs.OPEN_MEMORY) != 0 {
+		return file, flags, err
+	}
+
+	var cipher *xts.Cipher
+	if f, ok := name.DatabaseFile().(*xtsFile); ok {
+		cipher = f.cipher
+	} else {
+		var key []byte
+		if params := name.URIParameters(); name == nil {
+			key = x.init.KDF("") // Temporary files get a random key.
+		} else if t, ok := params["key"]; ok {
+			key = []byte(t[0])
+		} else if t, ok := params["hexkey"]; ok {
+			key, _ = hex.DecodeString(t[0])
+		} else if t, ok := params["textkey"]; ok {
+			key = x.init.KDF(t[0])
+		} else if flags&vfs.OPEN_MAIN_DB != 0 {
+			// Main datatabases may have their key specified as a PRAGMA.
+			return &xtsFile{File: file, init: x.init}, flags, nil
+		}
+		cipher = x.init.XTS(key)
+	}
+
+	if cipher == nil {
+		return nil, flags, sqlite3.CANTOPEN
+	}
+	return &xtsFile{File: file, cipher: cipher, init: x.init}, flags, nil
+}
+
+const sectorSize = 512
```

Comment:
Did you want to leave a comment that this isn't an XTS requirement, but doesn't seem to matter for performance?